primitives/src/assets.rs

```diff
@@ -0,0 +1,323 @@
+pub struct AssetStorageMetadata {
+	pub name: Vec<u8>,
+	pub symbol: Vec<u8>,
+	pub decimals: u8,
+	pub is_frozen: bool,
+}
+
+impl From<AssetRegistrarMetadata> for AssetStorageMetadata {
+	fn from(source: AssetRegistrarMetadata) -> Self {
+		AssetStorageMetadata {
+			name: source.name,
+			symbol: source.symbol,
+			decimals: source.decimals,
+			is_frozen: source.is_frozen,
+		}
+	}
+}
+
+#[derive(Clone, Eq, Debug, PartialEq, Encode, Decode, TypeInfo)]
+pub struct AssetLocation(pub VersionedMultiLocation);
+
+/// This cannot act as the default before v0.9.16 and need overwrite
+/// https://docs.google.com/document/d/1W8y00IcJb0JXPBF59aP4nm-c7DY8Ld02-yIAO7UxR80
+impl Default for AssetLocation {
+	fn default() -> Self {
+		AssetLocation(VersionedMultiLocation::V1(MultiLocation {
+			parents: 0,
+			interior: Junctions::Here,
+		}))
+	}
+}
+
+/// Convert a `MultiLocaiton` to an `AssetLocation`
+/// Note: This does not guaranttee the `AssetLocation` is registered (i.e. have an AssetId)
+impl From<MultiLocation> for AssetLocation {
+	fn from(location: MultiLocation) -> Self {
+		AssetLocation(VersionedMultiLocation::V1(location))
+	}
+}
+
+/// Convert an `AssetLocation` to a MultiLocation
+/// If Native, return none.
+impl From<AssetLocation> for Option<MultiLocation> {
+	fn from(location: AssetLocation) -> Self {
+		match location {
+			AssetLocation(VersionedMultiLocation::V1(location)) => Some(location),
+			_ => None,
+		}
+	}
+}
+
+/// Defines the trait to obtain a generic AssetId
+pub trait AssetIdLocationGetter<AssetLocation> {
+	// get AssetLocation from AssetId
+	fn get_asset_location(asset_id: AssetId) -> Option<AssetLocation>;
+
+	// get AssetId from AssetLocation
+	fn get_asset_id(loc: &AssetLocation) -> Option<AssetId>;
+}
+
+/// Defines the units per second charged given an `AssetId`.
+pub trait UnitsToWeightRatio {
+	/// Get units per second from asset id
+	fn get_units_per_second(asset_id: AssetId) -> Option<u128>;
+}
+
+/// Converter struct implementing `Convert`.
+/// This enforce the `AssetInfoGetter` implements `AssetIdLocationGetter`
+pub struct AssetIdLocationConvert<AssetLocation, AssetInfoGetter>(
+	PhantomData<(AssetLocation, AssetInfoGetter)>,
+);
+impl<AssetLocation, AssetInfoGetter> xcm_executor::traits::Convert<MultiLocation, AssetId>
+	for AssetIdLocationConvert<AssetLocation, AssetInfoGetter>
+where
+	AssetId: Clone,
+	AssetLocation: From<MultiLocation> + Into<Option<MultiLocation>> + Clone,
+	AssetInfoGetter: AssetIdLocationGetter<AssetLocation>,
+{
+	fn convert_ref(loc: impl Borrow<MultiLocation>) -> Result<AssetId, ()> {
+		AssetInfoGetter::get_asset_id(&loc.borrow().clone().into()).ok_or(())
+	}
+
+	fn reverse_ref(id: impl Borrow<AssetId>) -> Result<MultiLocation, ()> {
+		AssetInfoGetter::get_asset_location(id.borrow().clone())
+			.and_then(Into::into)
+			.ok_or(())
+	}
+}
+
+#[derive(Debug)]
+pub enum FungibleLedgerConsequence<Balance> {
```

Comment:
Is this ever used with a different type other than the (concrete) `Balance` type?

Reply:
We could concretize this as well.